DOC: sparse: Add migration guide for converting code from spmatrix to sparray (#20926)
scipy/scipy: doc/source/reference/sparse.migration_to_sparray.rst

.. _migration_to_sparray:

Migration from spmatrix to sparray
==================================

This document provides guidance for converting code from sparse matrices

> to sparse arrays in SciPy Sparse.

Reviewer: ```suggestion
to sparse arrays in ``scipy.sparse``.
```

> This document provides guidance for converting code from sparse matrices
> to sparse arrays in SciPy Sparse.

Reviewer: ```suggestion
This document provides guidance for converting code from sparse *matrices*
to sparse *arrays* in SciPy Sparse.
```

The change from sparse matrices to sparse arrays mirrors conversion from
``np.matrix`` to ``np.array``. Essentially we must move from an all-2D

> matrix multiplication centric ``matrix`` object to a 1D or 2D “array”

Reviewer: ```suggestion
matrix-multiplication-centric ``matrix`` object to a 1D or 2D “array”
```

> object that supports operand matrix multiplication and elementwise

Reviewer: ```suggestion
object that supports the matrix multiplication operator and elementwise
```

computation.

Notation: For this guide we denote the sparse array classes generally as
``sparray`` and the sparse matrix classes ``spmatrix``. Dense numpy
arrays are denoted ``np.array`` and dense matrix classes are
``np.matrix``. Supported sparse formats are denoted BSR, COO, CSC, CSR,
DIA, DOK, LIL and all formats are supported by both sparray and
spmatrix. The term ``sparse`` refers to either ``sparray`` or
``spmatrix``, while ``dense`` refers to either ``np.array`` or
``np.matrix``.

Overview and big picture:
-------------------------

-  the constructor names ``*_matrix``, e.g. ``csr_matrix``, are changed
   to ``*_array``.
-  spmatrix ``M`` is always 2D (rows x columns) even

>    e.g. ``M.min(axis=0)``. sparray ``A`` can be 1D or 2D. Numpy scalars

Reviewer: Perhaps missing something like:

> sparray preserves NumPy semantics, so that ``M.min(axis=0)`` returns a 1D result.

Reviewer: Not sure what it means to say "Numpy scalars are used as 0D values where needed."

Author: I included an example to make it more clear:
```
Numpy scalars are returned for full reductions, i.e. ``M.min()``.
```

   are used as 0D values where needed.
-  Operators change: ``*, @, *=, @=, **``

>    -  scalar multiplication uses ``*`` and is not implemented for ``@``.

Reviewer: Which scalars are we talking about here?

   -  sparrays use ``*`` for elementwise multiplication and ``@`` for
      matrix multiplication while spmatrices can use both operators

>       ``*`` and @ for matrix multiplication. Either can use

Reviewer: ```suggestion
      ``*`` and ``@`` for matrix multiplication. Either can use
```

      ``A.multiply(B)`` for elementwise multiplication.
   -  scalar exponents ``A**2`` use elementwise power for sparray and
      matrix power for spmatrix. To get matrix power for arrays use
      ``sp.sparse.linalg.matrix_power(A, n)``

-  Checking the sparse type and format:

   -  ``issparse(A)`` returns ``True`` for any sparse array/matrix.
   -  ``isspmatrix(M)`` returns ``True`` for any sparse matrix.
   -  ``issparse(A) and A.format == 'csr'`` looks for a CSR sparse
      array/matrix

Recommended steps for migration:
--------------------------------

-  First Pass (leaving spmatrix in the code)

   -  In your spmatrix code, change ``*`` to ``@`` for matrix
      multiplication. Note that scalar multiplication with sparse should
      use ``*``.
   -  scalar powers, e.g. ``M**3``, should be converted to
      ``sp.sparse.linalg.matrix_power(M, 3)`` or ``M @ M @ M`` if the

>       power is known.

Reviewer: Presumably the first call is faster than the second.

Author: I removed the slower multiple calls to the `@` operator to not confuse things.

Existing functions you should be careful while migrating
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

These functions return sparray or spmatrix dependent on their input. Use
of these should be examined and inputs adjusted to ensure return values
are sparrays. And in turn the outputs should be treated as sparrays.
To return sparrays, at least one input must be an sparray. If you use
list-of-lists or numpy arrays as input you should convert one of them
to a sparse array to get sparse arrays out.

::

   def kron(A, B, format=None):
   def kronsum(A, B, format=None):
   def hstack(blocks, format=None, dtype=None):
   def vstack(blocks, format=None, dtype=None):
   def block_diag(mats, format=None, dtype=None):
   def tril(A, k=0, format=None):
   def triu(A, k=0, format=None):

Functions that changed names for the migration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

   +=========+=============+
   |Function | New function|
   +=========+=============+
   |eye      | eye_array   |
   |identity | eye_array   |
   |diags    | diags_array |
   |spdiags  | diags_array |
   |bmat     | block       |

>    |rand     | random_arra |

Reviewer: typos

:func:`~scipy.sparse.hstack`, :func:`~scipy.sparse.vstack`,
:func:`~scipy.sparse.block_diag`, :func:`~scipy.sparse.tril`, and
:func:`~scipy.sparse.triu`. Their signatures are::

   def kron(A, B, format=None):
   def kronsum(A, B, format=None):
   def hstack(blocks, format=None, dtype=None):
   def vstack(blocks, format=None, dtype=None):
   def block_diag(mats, format=None, dtype=None):
   def tril(A, k=0, format=None):
   def triu(A, k=0, format=None):

Use of these functions should be examined and inputs adjusted to ensure return
values are sparrays. And in turn the outputs should be treated as sparrays.
To return sparrays, at least one input must be an sparray. If you use
list-of-lists or numpy arrays as input you should convert one of them
to a sparse array to get sparse arrays out.

Functions that changed names for the migration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

   +=========+=============+
   |Function | New function|
   +=========+=============+
   |eye      | eye_array   |
   |identity | eye_array   |
   |diags    | diags_array |
   |spdiags  | diags_array |
   |bmat     | block       |
   |rand     | random_array|
   |random   | random_array|

>    +=========+=============+

Reviewer 2: These tables don't render particularly nicely, but I'm happy to merge them - the perfect is the enemy of the good here I think.

Author: I changed the tables to rst simple table format in the latest commit. I like the ascii format better but this scales with browser size better. Let me know if you prefer to return to the ascii or suggest another way to make it work.
